exonum-java-binding/core/rust/src/proxy/runtime.rs

```diff
@@ -0,0 +1,596 @@
+
+    /// Creates new `JavaRuntimeProxy` for given `ServiceRuntimeAdapter` object
+    pub fn new(executor: Executor, adapter: GlobalRef) -> Self {
+        JavaRuntimeProxy {
+            exec: executor,
+            runtime_adapter: adapter,
+            deployed_artifacts: HashSet::new(),
+            started_services: HashMap::new(),
+            started_services_by_name: HashMap::new(),
+        }
+    }
+
+    fn parse_artifact(&self, artifact: &ArtifactId) -> Result<JavaArtifactId, ExecutionError> {
+        if artifact.runtime_id != Self::RUNTIME_ID as u32 {
+            return Err(Error::IncorrectArtifactId.into());
+        }
+        artifact
+            .name
+            .parse()
+            .map_err(|_| Error::IncorrectArtifactId.into())
+    }
+
+    fn add_started_service(&mut self, instance: Instance) {
+        self.started_services_by_name
+            .insert(instance.name.clone(), instance.id);
+        self.started_services.insert(instance.id, instance);
+    }
+
+    fn remove_started_service(&mut self, id: &InstanceId) {
+        let service = self.started_services.remove(id);
+        if let Some(instance) = service {
+            self.started_services_by_name.remove(&instance.name);
+        }
+    }
+
+    fn parse_jni<T>(res: JniResult<T>) -> Result<T, ExecutionError> {
+        res.map_err(|err| match err.0 {
+            JniErrorKind::JavaException => Error::JavaException.into(),
+            _ => Error::UnspecifiedError.into(),
+        })
+    }
+}
+
+impl fmt::Debug for JavaRuntimeProxy {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "JavaRuntimeProxy()")
+    }
+}
+
+impl From<Error> for ExecutionError {
+    fn from(value: Error) -> ExecutionError {
+        ExecutionError::new(
+            ErrorKind::runtime(value as u8),
+            format!("{:?}", value.clone())
+        )
+    }
+}
+
+impl Runtime for JavaRuntimeProxy {
+    fn deploy_artifact(
+        &mut self,
+        artifact: ArtifactId,
+        deploy_spec: Any
+    ) -> Box<dyn Future<Item = (), Error = ExecutionError>> {
+
+        let id = match self.parse_artifact(&artifact) {
+            Ok(id) => id,
+            Err(err) => return Box::new(Err(err).into_future()),
+        };
+
+        let execution_res = Self::parse_jni(self.exec.with_attached(|env| {
+            let artifact_id = JObject::from(
+                env.new_string(id.to_string())?
+            );
+            let spec = JObject::from(env.byte_array_from_slice(
+                &deploy_spec.into_bytes())?
+            );
+
+            env.call_method_unchecked(
+                self.runtime_adapter.as_obj(),
+                runtime_adapter::deploy_artifact_id(),
+                JavaType::from_str("V").unwrap(),
+                &[
+                    JValue::from(artifact_id),
+                    JValue::from(spec),
+                ],
+            )?;
+            Ok(())
+        }));
+
+        self.deployed_artifacts.insert(id);
```

Comment:
Why we insert id into a list of deployed artifacts even if an error returned from java?